packages/jaeger-ui/src/model/ddg/types.tsx

```diff
@@ -53,3 +55,56 @@ export type TDdgModel = {
 };
 
 export type TDdgVertex = TVertex<{ service: string; operation: string }>;
+
+export type TDdgStateEntry =
+  | {
+      state: typeof fetchedState.LOADING;
+    }
+  | {
+      error: ApiError;
+      state: typeof fetchedState.ERROR;
+    }
+  | {
+      model: TDdgModel;
+      state: typeof fetchedState.DONE;
+      viewModifiers: Map<number, number>;
+    };
+
+export const stateKey = ({ service, operation = '*', start, end }: TDdgModelParams): string =>
+  [service, operation, start, end].join('\t');
+
+export type TDdgState = Record<string, TDdgStateEntry>;
+
+export enum EViewModifier {
+  None,
+  Hovered,
+  Selected,
+  Emphasized = 1 << 2, // eslint-disable-line no-bitwise
+}
+
+export type TDdgModelParams = {
+  service: string;
+  operation?: string;
+  start: number;
+  end: number;
+};
+
+export type TDdgActionMeta = {
+  query: TDdgModelParams;
+};
+
+export type TDdgAddViewModifierPayload = TDdgModelParams & {
+  visibilityIndices: number[];
+  viewModifier: number;
+};
+
+export type TDdgClearViewModifiersFromIndicesPayload = TDdgAddViewModifierPayload & { viewModifier?: void };
```

Comment:
I don't see much benefit in basing these types off of `TDdgAddViewModifierPayload`.

Reply:
The reason I defined the types as `export type TDdgClearViewModifiersFromIndicesPayload = TDdgAddViewModifierPayload & { viewModifier?: void };` not `export type TDdgClearViewModifiersFromIndicesPayload = TDdgModelParams & { visibilityIndices: number [] };` Is because this line in the reducer `const { visibilityIndices, viewModifier } = payload;` would have the this type error: `Property 'viewModifier' does not exist on type 'TDdgViewModifierRemovalPayload'.`

I agree that the second type is a clearer, but typescript doesn't seem to like the implicit `number | undefined` when destructuring from the union type.
[Playground](https://www.typescriptlang.org/play/#code/C4TwDgpgBKkIJQLxQN4FgBQVtQHYEMBbCALigGdgAnAS1wHMBuTHKfAeVwBsQzcBXQgCMIVZhgC+mTLGiyAQklQscBYmUq0G41kM48+gkWMxSMmAGb9cAY2A0A9rhj4A1hHJx2VeQAow+FREZLIIAD4w4BDyAJTKWDg2TpSoeEQQADRs+iBZetwgUBJKAUGEOonJDlwQAHRcDvS+apnZBVAR+Twx4hJAA)